File: app/src/test/java/com/farmerbb/taskbar/content/TaskbarPositionTest.java
package com.farmerbb.taskbar.content;

import android.content.Context;
import android.view.Display;
import android.view.Surface;
import android.view.WindowManager;

import androidx.test.core.app.ApplicationProvider;

import com.farmerbb.taskbar.util.U;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.Shadows;
import org.robolectric.shadows.ShadowDisplay;

import java.util.ArrayList;
import java.util.List;

import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_BOTTOM_LEFT;
import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_BOTTOM_RIGHT;
import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_BOTTOM_VERTICAL_LEFT;
import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_BOTTOM_VERTICAL_RIGHT;
import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_TOP_LEFT;
import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_TOP_RIGHT;
import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_TOP_VERTICAL_LEFT;
import static com.farmerbb.taskbar.content.TaskbarPosition.POSITION_TOP_VERTICAL_RIGHT;
import static com.farmerbb.taskbar.content.TaskbarPosition.isBottom;
import static com.farmerbb.taskbar.content.TaskbarPosition.isLeft;
import static com.farmerbb.taskbar.content.TaskbarPosition.isRight;
import static com.farmerbb.taskbar.content.TaskbarPosition.isVertical;
import static com.farmerbb.taskbar.content.TaskbarPosition.isVerticalLeft;
import static com.farmerbb.taskbar.content.TaskbarPosition.isVerticalRight;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class TaskbarPositionTest {
    private static final String SP_KEY_ANCHOR = "anchor";
    private static final String SP_KEY_POSITION = "position";
    private Context context;

    @Before
    public void setUp() {
        context = ApplicationProvider.getApplicationContext();
        assertNotNull(context);
    }

    @Test
    public void testIsVertical() {
        assertFalse(isVertical(POSITION_BOTTOM_LEFT));
        assertFalse(isVertical(POSITION_BOTTOM_RIGHT));
        assertTrue(isVertical(POSITION_BOTTOM_VERTICAL_LEFT));
        assertTrue(isVertical(POSITION_BOTTOM_VERTICAL_RIGHT));
        assertFalse(isVertical(POSITION_TOP_LEFT));
        assertFalse(isVertical(POSITION_TOP_RIGHT));
        assertTrue(isVertical(POSITION_TOP_VERTICAL_LEFT));
        assertTrue(isVertical(POSITION_TOP_VERTICAL_RIGHT));
    }

    @Test
    public void testIsLeft() {
        assertTrue(isLeft(POSITION_BOTTOM_LEFT));
        assertFalse(isLeft(POSITION_BOTTOM_RIGHT));
        assertTrue(isLeft(POSITION_BOTTOM_VERTICAL_LEFT));
        assertFalse(isLeft(POSITION_BOTTOM_VERTICAL_RIGHT));
        assertTrue(isLeft(POSITION_TOP_LEFT));
        assertFalse(isLeft(POSITION_TOP_RIGHT));
        assertTrue(isLeft(POSITION_TOP_VERTICAL_LEFT));
        assertFalse(isLeft(POSITION_TOP_VERTICAL_RIGHT));
    }

    @Test
    public void testIsRight() {
        assertFalse(isRight(POSITION_BOTTOM_LEFT));
        assertTrue(isRight(POSITION_BOTTOM_RIGHT));
        assertFalse(isRight(POSITION_BOTTOM_VERTICAL_LEFT));
        assertTrue(isRight(POSITION_BOTTOM_VERTICAL_RIGHT));
        assertFalse(isRight(POSITION_TOP_LEFT));
        assertTrue(isRight(POSITION_TOP_RIGHT));
        assertFalse(isRight(POSITION_TOP_VERTICAL_LEFT));
        assertTrue(isRight(POSITION_TOP_VERTICAL_RIGHT));
    }

    @Test
    public void testIsBottom() {
        assertTrue(isBottom(POSITION_BOTTOM_LEFT));
        assertTrue(isBottom(POSITION_BOTTOM_RIGHT));
        assertTrue(isBottom(POSITION_BOTTOM_VERTICAL_LEFT));
        assertTrue(isBottom(POSITION_BOTTOM_VERTICAL_RIGHT));
        assertFalse(isBottom(POSITION_TOP_LEFT));
        assertFalse(isBottom(POSITION_TOP_RIGHT));
        assertFalse(isBottom(POSITION_TOP_VERTICAL_LEFT));
        assertFalse(isBottom(POSITION_TOP_VERTICAL_RIGHT));
    }

    @Test
    public void testIsVerticalLeft() {
        assertFalse(isVerticalLeft(POSITION_BOTTOM_LEFT));
        assertFalse(isVerticalLeft(POSITION_BOTTOM_RIGHT));
        assertTrue(isVerticalLeft(POSITION_BOTTOM_VERTICAL_LEFT));
        assertFalse(isVerticalLeft(POSITION_BOTTOM_VERTICAL_RIGHT));
        assertFalse(isVerticalLeft(POSITION_TOP_LEFT));
        assertFalse(isVerticalLeft(POSITION_TOP_RIGHT));
        assertTrue(isVerticalLeft(POSITION_TOP_VERTICAL_LEFT));
        assertFalse(isVerticalLeft(POSITION_TOP_VERTICAL_RIGHT));
    }

    @Test
    public void testIsVerticalRight() {
        assertFalse(isVerticalRight(POSITION_BOTTOM_LEFT));
        assertFalse(isVerticalRight(POSITION_BOTTOM_RIGHT));
        assertFalse(isVerticalRight(POSITION_BOTTOM_VERTICAL_LEFT));
        assertTrue(isVerticalRight(POSITION_BOTTOM_VERTICAL_RIGHT));
        assertFalse(isVerticalRight(POSITION_TOP_LEFT));
        assertFalse(isVerticalRight(POSITION_TOP_RIGHT));
        assertFalse(isVerticalRight(POSITION_TOP_VERTICAL_LEFT));
        assertTrue(isVerticalRight(POSITION_TOP_VERTICAL_RIGHT));
    }

    @Test
    public void testGetTaskbarPositionWithoutAnchor() {
        String position = TaskbarPosition.getTaskbarPosition(context);
        // The default position is bottom_left
        assertEquals(POSITION_BOTTOM_LEFT, position);
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionBottomLeft() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_BOTTOM_LEFT,
                new ArrayList<String>() {{
                    add(POSITION_BOTTOM_LEFT);
                    add(POSITION_BOTTOM_VERTICAL_RIGHT);
                    add(POSITION_TOP_RIGHT);
                    add(POSITION_TOP_VERTICAL_LEFT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionBottomVerticalLeft() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_BOTTOM_VERTICAL_LEFT,
                new ArrayList<String>() {{
                    add(POSITION_BOTTOM_VERTICAL_LEFT);
                    add(POSITION_BOTTOM_RIGHT);
                    add(POSITION_TOP_VERTICAL_RIGHT);
                    add(POSITION_TOP_LEFT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionBottomRight() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_BOTTOM_RIGHT,
                new ArrayList<String>() {{
                    add(POSITION_BOTTOM_RIGHT);
                    add(POSITION_TOP_VERTICAL_RIGHT);
                    add(POSITION_TOP_LEFT);
                    add(POSITION_BOTTOM_VERTICAL_LEFT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionBottomVerticalRight() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_BOTTOM_VERTICAL_RIGHT,
                new ArrayList<String>() {{
                    add(POSITION_BOTTOM_VERTICAL_RIGHT);
                    add(POSITION_TOP_RIGHT);
                    add(POSITION_TOP_VERTICAL_LEFT);
                    add(POSITION_BOTTOM_LEFT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionTopLeft() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_TOP_LEFT,
                new ArrayList<String>() {{
                    add(POSITION_TOP_LEFT);
                    add(POSITION_BOTTOM_VERTICAL_LEFT);
                    add(POSITION_BOTTOM_RIGHT);
                    add(POSITION_TOP_VERTICAL_RIGHT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionTopVerticalLeft() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_TOP_VERTICAL_LEFT,
                new ArrayList<String>() {{
                    add(POSITION_TOP_VERTICAL_LEFT);
                    add(POSITION_BOTTOM_LEFT);
                    add(POSITION_BOTTOM_VERTICAL_RIGHT);
                    add(POSITION_TOP_RIGHT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionTopRight() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_TOP_RIGHT,
                new ArrayList<String>() {{
                    add(POSITION_TOP_RIGHT);
                    add(POSITION_TOP_VERTICAL_LEFT);
                    add(POSITION_BOTTOM_LEFT);
                    add(POSITION_BOTTOM_VERTICAL_RIGHT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndPositionTopVerticalRight() {
        checkTaskbarPositionWithDifferentRotation(
                POSITION_TOP_VERTICAL_RIGHT,
                new ArrayList<String>() {{
                    add(POSITION_TOP_VERTICAL_RIGHT);
                    add(POSITION_TOP_LEFT);
                    add(POSITION_BOTTOM_VERTICAL_LEFT);
                    add(POSITION_BOTTOM_RIGHT);
                }}
        );
    }

    @Test
    public void testGetTaskbarPositionWithAnchorAndInvalidPosition() {
        checkTaskbarPositionWithDifferentRotation(
                "invalid-position",
                new ArrayList<String>() {{
                    add(POSITION_BOTTOM_LEFT);
                    add(POSITION_BOTTOM_LEFT);
                    add(POSITION_BOTTOM_LEFT);
                    add(POSITION_BOTTOM_LEFT);
                }}
        );
    }

    private void checkTaskbarPositionWithDifferentRotation(String originPosition,
                                                           List<String> changedPositions) {
        assertEquals(4, changedPositions.size());
        String oldPosition =
                U.getSharedPreferences(context).getString(SP_KEY_POSITION, POSITION_BOTTOM_LEFT);
        boolean oldAnchor =
                U.getSharedPreferences(context).getBoolean(SP_KEY_ANCHOR, false);
        initializeTaskbarPosition(originPosition);
        initializeRotation(Surface.ROTATION_0);
        assertEquals(changedPositions.get(0), TaskbarPosition.getTaskbarPosition(context));
        initializeRotation(Surface.ROTATION_90);
        assertEquals(changedPositions.get(1), TaskbarPosition.getTaskbarPosition(context));
        initializeRotation(Surface.ROTATION_180);
        assertEquals(changedPositions.get(2), TaskbarPosition.getTaskbarPosition(context));
        initializeRotation(Surface.ROTATION_270);
        assertEquals(changedPositions.get(3), TaskbarPosition.getTaskbarPosition(context));
        U.getSharedPreferences(context).edit().putBoolean(SP_KEY_ANCHOR, oldAnchor).apply();
        U.getSharedPreferences(context).edit().putString(SP_KEY_POSITION, oldPosition).apply();
    }

    private void initializeTaskbarPosition(String position) {
        U.getSharedPreferences(context).edit().putBoolean(SP_KEY_ANCHOR, true).apply();
        U.getSharedPreferences(context).edit().putString(SP_KEY_POSITION, position).apply();
    }

    private void initializeRotation(int rotation) {
        WindowManager windowManager =
                (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = windowManager.getDefaultDisplay();
        ShadowDisplay shadowDisplay = Shadows.shadowOf(display);
        shadowDisplay.setRotation(rotation);
    }
}

File: app/src/test/java/com/farmerbb/taskbar/util/TaskbarPositionTest.kt
package com.farmerbb.taskbar.util

import android.content.Context
import android.view.Surface
import android.view.WindowManager
import androidx.test.core.app.ApplicationProvider
import java.util.function.Predicate
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows

@RunWith(RobolectricTestRunner::class)
class TaskbarPositionTest {
    private lateinit var context: Context

    @Before
    fun setUp() {
        context = ApplicationProvider.getApplicationContext()
        Assert.assertNotNull(context)
    }

    @Test
    fun testIsVertical() {
        Assert.assertFalse(TaskbarPosition.isVertical(Constants.POSITION_BOTTOM_LEFT))
        Assert.assertFalse(TaskbarPosition.isVertical(Constants.POSITION_BOTTOM_RIGHT))
        Assert.assertTrue(TaskbarPosition.isVertical(Constants.POSITION_BOTTOM_VERTICAL_LEFT))
        Assert.assertTrue(TaskbarPosition.isVertical(Constants.POSITION_BOTTOM_VERTICAL_RIGHT))
        Assert.assertFalse(TaskbarPosition.isVertical(Constants.POSITION_TOP_LEFT))
        Assert.assertFalse(TaskbarPosition.isVertical(Constants.POSITION_TOP_RIGHT))
        Assert.assertTrue(TaskbarPosition.isVertical(Constants.POSITION_TOP_VERTICAL_LEFT))
        Assert.assertTrue(TaskbarPosition.isVertical(Constants.POSITION_TOP_VERTICAL_RIGHT))
    }

    @Test
    fun testIsVerticalWithContext() {
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_LEFT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_RIGHT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_LEFT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_RIGHT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isVertical(context) },
                true
        )
    }

    @Test
    fun testIsLeft() {
        Assert.assertTrue(TaskbarPosition.isLeft(Constants.POSITION_BOTTOM_LEFT))
        Assert.assertFalse(TaskbarPosition.isLeft(Constants.POSITION_BOTTOM_RIGHT))
        Assert.assertTrue(TaskbarPosition.isLeft(Constants.POSITION_BOTTOM_VERTICAL_LEFT))
        Assert.assertFalse(TaskbarPosition.isLeft(Constants.POSITION_BOTTOM_VERTICAL_RIGHT))
        Assert.assertTrue(TaskbarPosition.isLeft(Constants.POSITION_TOP_LEFT))
        Assert.assertFalse(TaskbarPosition.isLeft(Constants.POSITION_TOP_RIGHT))
        Assert.assertTrue(TaskbarPosition.isLeft(Constants.POSITION_TOP_VERTICAL_LEFT))
        Assert.assertFalse(TaskbarPosition.isLeft(Constants.POSITION_TOP_VERTICAL_RIGHT))
    }

    @Test
    fun testIsLeftWithContext() {
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_LEFT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_RIGHT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_LEFT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_RIGHT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isLeft(context) },
                false
        )
    }

    @Test
    fun testIsRight() {
        Assert.assertFalse(TaskbarPosition.isRight(Constants.POSITION_BOTTOM_LEFT))
        Assert.assertTrue(TaskbarPosition.isRight(Constants.POSITION_BOTTOM_RIGHT))
        Assert.assertFalse(TaskbarPosition.isRight(Constants.POSITION_BOTTOM_VERTICAL_LEFT))
        Assert.assertTrue(TaskbarPosition.isRight(Constants.POSITION_BOTTOM_VERTICAL_RIGHT))
        Assert.assertFalse(TaskbarPosition.isRight(Constants.POSITION_TOP_LEFT))
        Assert.assertTrue(TaskbarPosition.isRight(Constants.POSITION_TOP_RIGHT))
        Assert.assertFalse(TaskbarPosition.isRight(Constants.POSITION_TOP_VERTICAL_LEFT))
        Assert.assertTrue(TaskbarPosition.isRight(Constants.POSITION_TOP_VERTICAL_RIGHT))
    }

    @Test
    fun testIsRightWithContext() {
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_LEFT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_RIGHT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_LEFT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_RIGHT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isRight(context) },
                true
        )
    }

    @Test
    fun testIsBottom() {
        Assert.assertTrue(TaskbarPosition.isBottom(Constants.POSITION_BOTTOM_LEFT))
        Assert.assertTrue(TaskbarPosition.isBottom(Constants.POSITION_BOTTOM_RIGHT))
        Assert.assertTrue(TaskbarPosition.isBottom(Constants.POSITION_BOTTOM_VERTICAL_LEFT))
        Assert.assertTrue(TaskbarPosition.isBottom(Constants.POSITION_BOTTOM_VERTICAL_RIGHT))
        Assert.assertFalse(TaskbarPosition.isBottom(Constants.POSITION_TOP_LEFT))
        Assert.assertFalse(TaskbarPosition.isBottom(Constants.POSITION_TOP_RIGHT))
        Assert.assertFalse(TaskbarPosition.isBottom(Constants.POSITION_TOP_VERTICAL_LEFT))
        Assert.assertFalse(TaskbarPosition.isBottom(Constants.POSITION_TOP_VERTICAL_RIGHT))
    }

    @Test
    fun testIsBottomWithContext() {
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_LEFT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_RIGHT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_LEFT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_RIGHT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isBottom(context) },
                false
        )
    }

    @Test
    fun testIsVerticalLeft() {
        Assert.assertFalse(TaskbarPosition.isVerticalLeft(Constants.POSITION_BOTTOM_LEFT))
        Assert.assertFalse(TaskbarPosition.isVerticalLeft(Constants.POSITION_BOTTOM_RIGHT))
        Assert.assertTrue(TaskbarPosition.isVerticalLeft(Constants.POSITION_BOTTOM_VERTICAL_LEFT))
        Assert.assertFalse(TaskbarPosition.isVerticalLeft(Constants.POSITION_BOTTOM_VERTICAL_RIGHT))
        Assert.assertFalse(TaskbarPosition.isVerticalLeft(Constants.POSITION_TOP_LEFT))
        Assert.assertFalse(TaskbarPosition.isVerticalLeft(Constants.POSITION_TOP_RIGHT))
        Assert.assertTrue(TaskbarPosition.isVerticalLeft(Constants.POSITION_TOP_VERTICAL_LEFT))
        Assert.assertFalse(TaskbarPosition.isVerticalLeft(Constants.POSITION_TOP_VERTICAL_RIGHT))
    }

    @Test
    fun testIsVerticalLeftWithContext() {
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalLeft(context) },
                false
        )
    }

    @Test
    fun testIsVerticalRight() {
        Assert.assertFalse(TaskbarPosition.isVerticalRight(Constants.POSITION_BOTTOM_LEFT))
        Assert.assertFalse(TaskbarPosition.isVerticalRight(Constants.POSITION_BOTTOM_RIGHT))
        Assert.assertFalse(TaskbarPosition.isVerticalRight(Constants.POSITION_BOTTOM_VERTICAL_LEFT))
        Assert.assertTrue(TaskbarPosition.isVerticalRight(Constants.POSITION_BOTTOM_VERTICAL_RIGHT))
        Assert.assertFalse(TaskbarPosition.isVerticalRight(Constants.POSITION_TOP_LEFT))
        Assert.assertFalse(TaskbarPosition.isVerticalRight(Constants.POSITION_TOP_RIGHT))
        Assert.assertFalse(TaskbarPosition.isVerticalRight(Constants.POSITION_TOP_VERTICAL_LEFT))
        Assert.assertTrue(TaskbarPosition.isVerticalRight(Constants.POSITION_TOP_VERTICAL_RIGHT))
    }

    @Test
    fun testIsVerticalRightWithContext() {
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_BOTTOM_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                true
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_LEFT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                false
        )
        checkTaskbarPositionGroup(
                Constants.POSITION_TOP_VERTICAL_RIGHT,
                { context: Context? -> TaskbarPosition.isVerticalRight(context) },
                true
        )
    }

    @Test
    fun testGetTaskbarPositionWithoutAnchor() {
        val position = TaskbarPosition.getTaskbarPosition(context)
        // The default position is bottom_left
        Assert.assertEquals(Constants.POSITION_BOTTOM_LEFT, position)
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionBottomLeft() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_BOTTOM_LEFT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_BOTTOM_LEFT)
                        add(Constants.POSITION_BOTTOM_VERTICAL_RIGHT)
                        add(Constants.POSITION_TOP_RIGHT)
                        add(Constants.POSITION_TOP_VERTICAL_LEFT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionBottomVerticalLeft() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_BOTTOM_VERTICAL_LEFT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_BOTTOM_VERTICAL_LEFT)
                        add(Constants.POSITION_BOTTOM_RIGHT)
                        add(Constants.POSITION_TOP_VERTICAL_RIGHT)
                        add(Constants.POSITION_TOP_LEFT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionBottomRight() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_BOTTOM_RIGHT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_BOTTOM_RIGHT)
                        add(Constants.POSITION_TOP_VERTICAL_RIGHT)
                        add(Constants.POSITION_TOP_LEFT)
                        add(Constants.POSITION_BOTTOM_VERTICAL_LEFT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionBottomVerticalRight() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_BOTTOM_VERTICAL_RIGHT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_BOTTOM_VERTICAL_RIGHT)
                        add(Constants.POSITION_TOP_RIGHT)
                        add(Constants.POSITION_TOP_VERTICAL_LEFT)
                        add(Constants.POSITION_BOTTOM_LEFT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionTopLeft() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_TOP_LEFT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_TOP_LEFT)
                        add(Constants.POSITION_BOTTOM_VERTICAL_LEFT)
                        add(Constants.POSITION_BOTTOM_RIGHT)
                        add(Constants.POSITION_TOP_VERTICAL_RIGHT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionTopVerticalLeft() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_TOP_VERTICAL_LEFT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_TOP_VERTICAL_LEFT)
                        add(Constants.POSITION_BOTTOM_LEFT)
                        add(Constants.POSITION_BOTTOM_VERTICAL_RIGHT)
                        add(Constants.POSITION_TOP_RIGHT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionTopRight() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_TOP_RIGHT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_TOP_RIGHT)
                        add(Constants.POSITION_TOP_VERTICAL_LEFT)
                        add(Constants.POSITION_BOTTOM_LEFT)
                        add(Constants.POSITION_BOTTOM_VERTICAL_RIGHT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndPositionTopVerticalRight() {
        checkTaskbarPositionWithDifferentRotation(
                Constants.POSITION_TOP_VERTICAL_RIGHT,
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_TOP_VERTICAL_RIGHT)
                        add(Constants.POSITION_TOP_LEFT)
                        add(Constants.POSITION_BOTTOM_VERTICAL_LEFT)
                        add(Constants.POSITION_BOTTOM_RIGHT)
                    }
                }
        )
    }

    @Test
    fun testGetTaskbarPositionWithAnchorAndInvalidPosition() {
        checkTaskbarPositionWithDifferentRotation(
                "invalid-position",
                object : ArrayList<String>() {
                    init {
                        add(Constants.POSITION_BOTTOM_LEFT)
                        add(Constants.POSITION_BOTTOM_LEFT)
                        add(Constants.POSITION_BOTTOM_LEFT)
                        add(Constants.POSITION_BOTTOM_LEFT)
                    }
                }
        )
    }

    private fun checkTaskbarPositionWithDifferentRotation(
        originPosition: String,
        changedPositions: List<String>
    ) {
        Assert.assertEquals(4, changedPositions.size.toLong())
        val oldPosition = U.getSharedPreferences(context)
                .getString(Constants.PREF_POSITION, Constants.POSITION_BOTTOM_LEFT)
        val oldAnchor = U.getSharedPreferences(context)
                .getBoolean(Constants.PREF_ANCHOR, false)
        initializeTaskbarPosition(originPosition)
        initializeRotation(Surface.ROTATION_0)
        Assert.assertEquals(changedPositions[0], TaskbarPosition.getTaskbarPosition(context))
        initializeRotation(Surface.ROTATION_90)
        Assert.assertEquals(changedPositions[1], TaskbarPosition.getTaskbarPosition(context))
        initializeRotation(Surface.ROTATION_180)
        Assert.assertEquals(changedPositions[2], TaskbarPosition.getTaskbarPosition(context))
        initializeRotation(Surface.ROTATION_270)
        Assert.assertEquals(changedPositions[3], TaskbarPosition.getTaskbarPosition(context))
        U.getSharedPreferences(context).edit()
                .putBoolean(Constants.PREF_ANCHOR, oldAnchor).apply()
        U.getSharedPreferences(context).edit()
                .putString(Constants.PREF_POSITION, oldPosition).apply()
    }

    private fun checkTaskbarPositionGroup(
        originPosition: String,
        predicate: Predicate<Context?>,
        expectedResult: Boolean
    ) {
        val oldPosition = U.getSharedPreferences(context)
                .getString(Constants.PREF_POSITION, Constants.POSITION_BOTTOM_LEFT)
        val oldAnchor = U.getSharedPreferences(context)
                .getBoolean(Constants.PREF_ANCHOR, false)
        initializeTaskbarPosition(originPosition)
        Assert.assertEquals(expectedResult, predicate.test(context))
        U.getSharedPreferences(context).edit()
                .putBoolean(Constants.PREF_ANCHOR, oldAnchor).apply()
        U.getSharedPreferences(context).edit()
                .putString(Constants.PREF_POSITION, oldPosition).apply()
    }

    private fun initializeTaskbarPosition(position: String) {
        U.getSharedPreferences(context).edit()
                .putBoolean(Constants.PREF_ANCHOR, true).apply()
        U.getSharedPreferences(context).edit()
                .putString(Constants.PREF_POSITION, position).apply()
    }

    private fun initializeRotation(rotation: Int) {
        val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = windowManager.defaultDisplay
        val shadowDisplay = Shadows.shadowOf(display)
        shadowDisplay.setRotation(rotation)
    }
}
